bi-connectors/PowerBIConnector/src/OpenSearchProject.pq

```diff
@@ -207,56 +225,174 @@ OpenSearchProject = [
+                ]
+            ),
+            Parts = Json.Document(Response)
+        in
+            Parts,
+
+    Refresh = (refreshToken) =>
+        let
+            tokenUrl = TokenUrl,
+            requestBody = "grant_type=refresh_token&refresh_token=" & refreshToken & "&client_id=" & ClientId & "&client_secret=" & ClientSecret,
+            refreshResponse = Json.Document(Web.Contents(tokenUrl, [Content = Text.ToBinary(requestBody), Headers = [#"Content-Type"="application/x-www-form-urlencoded"]])),
+            refreshedAccessToken = refreshResponse[access_token]
+        in
+            refreshedAccessToken,
+
+    Logout = (accessToken) =>
+        let
+            logoutUrl = LogoutUrl,
+            redirectUrl = callbackUri,
+            response = Web.Contents(logoutUrl, [Query = [redirect_uri = redirectUrl], Headers = [Authorization="Bearer " & accessToken]])
+        in
+            // Assuming successful logout
+            null,
+
     // PBIDS Handler
     DSRHandlers = [
-      opensearchproject = [
-        GetDSR = (Server, Port, UseSSL, HostnameVerification, optional Options) => [ protocol = "opensearchproject-odbc", address = [ server = Server, port = Port, useSSL = UseSSL, hostnameVerification = HostnameVerification ] ],
-        GetFormula = (dsr, optional options) => () =>
-          let
-            db = OpenSearchProject.Contents(dsr[address][server], dsr[address][port], dsr[address][useSSL], dsr[address][hostnameVerification])
-          in
-            db,
-        GetFriendlyName = (dsr) => "OpenSearch Project"
-      ]
+        opensearchproject = [
+            GetDSR = (Server, Port, UseSSL, HostnameVerification, IDPurl, ClientId, ClientSecret, optional Options) =>
+                [
+                    protocol = "opensearchproject-odbc",
+                    address = [
+                        server = Server,
+                        port = Port,
+                        useSSL = UseSSL,
+                        idpURL = IDPurl,
+                        username = ClientId,
+                        secret = ClientSecret,
+                        hostnameVerification = HostnameVerification
+                    ]
+                ],
+            GetFormula = (dsr, optional options) =>
+                () =>
+                    let
+                        db = OpenSearchProject.Contents(
+                            dsr[address][server],
+                            dsr[address][port],
+                            dsr[address][useSSL],
+                            dsr[address][idpURL],
+                            dsr[address][ClientId],
+                            dsr[address][ClientSecret],                            
+                            dsr[address][hostnameVerification]
+                        )
+                    in
+                        db,
+            GetFriendlyName = (dsr) => "OpenSearch Projectatr"
```

Comment:
Is name change to Projectatr intentional?

Reply:
No, is typo. I will correct.